### PR TITLE
Make fem::Form stateless

### LINKED ISSUE #1164: Create fem::Form and fem::FormIntegral objects fully at construction
`fem::Form` and `fem::FormIntegral` have 'set' functions, which is a highly undesirable design as it means with have objects in some semi-complete state of construction. We're striving to remove this design pattern in DOLFIN-X.

If something is really needed for the Python bindings we should mark it as such and warn against using it in C++ code.

### PR DESCRIPTION
**Pull request by garth-wells**
The change fully creates a usable Form object at construction.

The old design was an anti-pattern with incomplete construction, even requiring a class method to complete construction. Changes used maps rather than array with mapping arrays, which simplifies the code. 

There are more possible changes to simplify the code, but the PR at least largely encapsulates them by making Form stateless. 

Fixes #1164.

### CODE REVIEW

File: python/test/unit/fem/test_custom_jit_kernels.py [hunk 1]
```diff
@@ -114,7 +113,7 @@ def test_coefficient():
     assert (np.isclose(bnorm, 2.0 * 0.0739710713711999))
 
 
-@skip_if_complex
+@ skip_if_complex
```
Comment (francesco-ballarin) [inline python/test/unit/fem/test_custom_jit_kernels.py:116]: were these space changes (here, and the two below) done on purpose or are they an artifact of the text editor?

File: cpp/dolfinx/fem/Form.h [hunk 2]
```diff
@@ -1,19 +1,16 @@
-// Copyright (C) 2007-2014 Anders Logg
+// Copyright (C) 2019-200 Garth N. Wells and Chris Richardson
```
Comment (francesco-ballarin) [inline cpp/dolfinx/fem/Form.h:1]: 2020

Comment (garth-wells) [reply cpp/dolfinx/fem/Form.h]: Fixed

File: python/test/unit/fem/test_custom_jit_kernels.py (hunk 1 above)
Comment (garth-wells) [reply python/test/unit/fem/test_custom_jit_kernels.py]: Weird - must gave been an artifact of the text editor. Fixed now.

Comment (francesco-ballarin) [reply python/test/unit/fem/test_custom_jit_kernels.py]: do you want to do the same white space fix for the two includes?

Comment (garth-wells) [reply python/test/unit/fem/test_custom_jit_kernels.py]: Where are includes?

File: python/test/unit/fem/test_custom_jit_kernels.py [hunk 3]
```diff
@@ -123,8 +122,8 @@ def test_cffi_assembly():
         from cffi import FFI
         ffibuilder = FFI()
         ffibuilder.set_source("_cffi_kernelA", r"""
-        #include <math.h>
-        #include <stdalign.h>
+        # include <math.h>
```
Comment (francesco-ballarin) [inline python/test/unit/fem/test_custom_jit_kernels.py:125]: possibly same editor issue (here and on the next line)

File: cpp/dolfinx/fem/Form.h [hunk 4]
```diff
@@ -214,67 +167,322 @@ class Form
...
-  /// Register the function for 'tabulate_tensor' for cell integral i
-  void set_tabulate_tensor(
-      IntegralType type, int i,
-      const std::function<void(T*, const T*, const T*, const double*,
-                               const int*, const std::uint8_t*,
-                               const std::uint32_t)>& fn)
+  /// Get the function for 'kernel' for integral i of given
+  /// type
+  /// @param[in] type Integral type
+  /// @param[in] i Integral number
```
Comment (michalhabera) [inline cpp/dolfinx/fem/Form.h:173]: Do you mean the integral ID here? I.e. the ID of the domain the integral is defined on.

File: cpp/dolfinx/fem/utils.h [hunk 5]
```diff
@@ -223,60 +214,141 @@ Form<T> create_form(
...
+  if (const_name.size() != constants.size())
+    throw std::runtime_error("Too few constants for form.");
+
+  // Place coefficients in appropriate order
+  std::vector<std::shared_ptr<const function::Constant<T>>> const_map(
+      ufc_form.num_constants);
+  for (auto& c : constants)
+  {
+    auto it = std::find(const_name.begin(), const_name.end(), c.first);
+    if (it == coeff_name.end())
```
Comment (michalhabera) [inline cpp/dolfinx/fem/utils.h:316]: const_name?

**Review by michalhabera — COMMENTED** [review]
I like the changes.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 6]
```diff
@@ -171,7 +171,7 @@ void _lift_bc_cells(
 
   const std::function<void(T*, const T*, const T*, const double*, const int*,
                            const std::uint8_t*, const std::uint32_t)>& fn
-      = a.integrals().get_tabulate_tensor(IntegralType::cell, 0);
+      = a.kernel(IntegralType::cell, -1);
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_vector_impl.h:174]: The is an existing bug - see https://github.com/FEniCS/dolfinx/issues/1110.

File: cpp/dolfinx/fem/Form.h [hunk 7]
```diff
@@ -214,67 +167,322 @@ class Form
...
   {
     return _constants;
   }
 
 private:
-  // Integrals associated with the Form
-  FormIntegrals<T> _integrals;
+  /// Set the valid domains for the integrals of a given type from a
+  /// MeshTags "marker". Note the MeshTags is not stored, so if there
+  /// any changes to the integration domain this must be called again.
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/Form.h]: Create a new form . . .

File: cpp/dolfinx/fem/Form.h [hunk 8]
```diff
@@ -214,67 +167,322 @@ class Form
...
+    }
 
-  // Constants associated with the Form
-  std::vector<
-      std::pair<std::string, std::shared_ptr<const function::Constant<T>>>>
-      _constants;
+    auto it0 = _integrals.find(type);
+    if (it0 == _integrals.end())
+    {
+      // TODO: No kernels for domain of this dimension - Add warning
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/Form.h]: This case can't arise in the new design, so I've now added an assertion.

File: cpp/dolfinx/fem/Form.h [hunk 9]
```diff
@@ -214,67 +167,322 @@ class Form
...
+  std::vector<std::shared_ptr<const function::Function<T>>> _coefficients;
+
+  // Constants associated with the Form
+  std::vector<std::shared_ptr<const function::Constant<T>>> _constants;
+
+  // The mesh
   std::shared_ptr<const mesh::Mesh> _mesh;
-};
+
+  using kern
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/Form.h:473]: Let's discuss the separately. I often find that a heavily used typedef declared far from where it's used can reduce readability.

File: cpp/dolfinx/fem/utils.h (hunk 5 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/utils.h]: Fixed.

File: cpp/dolfinx/fem/Form.h (hunk 4 above)
Comment (garth-wells) [reply cpp/dolfinx/fem/Form.h:173]: Yes, docs fixed.

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalhabera) [reply cpp/dolfinx/fem/assemble_vector_impl.h:174]: Ouch... Should be fixed with priority.

**Pull request merged**